SI-6502 Reenables loading jars into the running REPL (regression in 2.10)

heathermiller opened the pull request:
Fixes [SI-6502](https://issues.scala-lang.org/browse/SI-6502), reenables loading jars into the running REPL (regression in 2.10). This PR allows adding a jar to the compile and runtime classpaths without resetting the REPL state (crucial for Spark [SPARK-3257](issues.apache.org/jira/browse/SPARK-3257)).

This follows the lead taken by @som-snytt in PR [#3986](https://github.com/scala/scala/pull/3986), which differentiates two jar-loading behaviors (muddled by `cp`):
- adding jars and replaying REPL expressions (using `replay`)
- adding jars without resetting the REPL (deprecated `cp`, introduced `require`)
  This PR implements `require` (left unimplemented in [#3986](https://github.com/scala/scala/pull/3986))

This PR is a simplification of a similar approach taken by @gkossakowski in #3884. In this attempt, we check first to make sure that a jar is only added if it only contains _new_ classes/traits/objects, otherwise we emit an error. This differs from the old invalidation approach which also tracked deleted classpath entries.

heathermiller committed:
SI-6502 Reenables loading jars into the running REPL (regression in 2.10)

Fixes SI-6502, reenables loading jars into the running REPL
(regression in 2.10). This PR allows adding a jar to the compile
and runtime classpaths without resetting the REPL state (crucial
for Spark SPARK-3257).

This follows the lead taken by @som-snytt in PR #3986, which
differentiates two jar-loading behaviors (muddled by cp):

  - adding jars and replaying REPL expressions (using replay)
  - adding jars without resetting the REPL (deprecated cp,
    introduced require) This PR implements require (left
    unimplemented in #3986)

This PR is a simplification of a similar approach taken by
@gkossakowski in #3884. In this attempt, we check first to make
sure that a jar is only added if it only contains new
classes/traits/objects, otherwise we emit an error. This differs
from the old invalidation approach which also tracked deleted
classpath entries.

heathermiller commented:
Review by anyone interested. Especially those I pummeled with questions: @dragos, @phaller, @gkossakowski

retronym commented:
It is probably best to submit a squashed PR and point reviewers to the fine grained commits on a branch. That means less work for our build :cat: :)

retronym commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -842,6 +842,147 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
+   *  @return A pair consisting of
+   *    - a list of invalidated packages
+   *    - a list of of packages that should have been invalidated but were not because
+   *      they are system packages.
```
Rather than documenting this, how about encapsulating the return values in a case class?

heathermiller replied:
Actually, there’s no need for return values, I realize. So this now returns Unit.

retronym commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -606,13 +606,73 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+        val arr = readFully(is)
+        val clazz = cloader.classOf(arr)
+        try {
+          Class.forName(clazz.getName(), false, intp.classLoader)
```
Should document why we pass `initialize = false` here ("we don't want to execute class initializers")

som-snytt replied:
`intp.classLoader.tryToLoadClass` contrasts with `tryToInitializeClass`, in `ScalaClassLoader`.

heathermiller replied:
Done 'n documented, thanks!

retronym commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -606,13 +606,73 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+    while (entries.hasMoreElements()) {
+      val entry = entries.nextElement()
+      // skip directories and manifests
+      if (!entry.isDirectory() && !entry.getName().endsWith("MF")) {
```
What about other resources in the JAR (e.g. `.properties`)? I suppose the code below will work out okay, but it seems cleaner invert this condition to `entry.isFile && entry.getName.endsWith(".class")`.

heathermiller replied:
Thanks, done!

retronym commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -606,13 +606,73 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
-      echo("Added '%s'.  Your new classpath is:\n\"%s\"".format(f.path, totalClasspath))
-      replay()
+      intp.addUrlsToClassPath(f.toURI.toURL)
+      echo("Added '%s'.  Your new classpath is:\n\"%s\"".format(f.path, intp.global.classPath.asClasspathString))
```
This might be a scary wall of text that most users don't want to see, maybe demote the full classpath to debug logging.

heathermiller replied:
Sure thing. Spark spits out the whole classpath, so that’s why I kept it. But I’ve now got it echoing just what’s been added, and the full classpath is now left for debug.

retronym commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -221,7 +221,7 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
     nullary("quit", "exit the interpreter", () => Result(keepRunning = false, None)),
     cmd("replay", "[options]", "reset the repl and replay all previous commands", replayCommand),
-    //cmd("require", "<path>", "add a jar or directory to the classpath", require),  // TODO
+    cmd("require", "<path>", "add a jar or directory to the classpath", require),
```
"add a jar to the classpath".

heathermiller replied:
Done :), thanks for catching this!

heathermiller commented:
commits squashed :)

retronym commented:
I'd like to see this built on top of our existing abstractions for JAR file handling (`AbstractFile`).

``` scala
scala> def flatten(f: AbstractFile): Iterator[AbstractFile] = if (f.isClassContainer) f.iterator.flatMap(flatten) else Iterator(f)
flatten: (f: scala.reflect.io.AbstractFile)Iterator[scala.reflect.io.AbstractFile]

scala> val jar = AbstractFile.getDirectory(new java.io.File("/code/scala/build/pack/lib/scala-library.jar"))
jar: scala.reflect.io.AbstractFile = /code/scala/build/pack/lib/scala-library.jar

scala> flatten(jar).map(_.path).take(15).mkString("\n")
res7: String =
rootdoc.txt
META-INF/MANIFEST.MF
library.properties
scala/deprecatedName.class
scala/Function1$mcVI$sp$class.class
scala/Function0$mcJ$sp$class.class
scala/Predef$StringAdd$.class
scala/Function2$mcDID$sp$class.class
scala/Function2$mcJJD$sp.class
scala/Product2$mcJD$sp$class.class
scala/Product2$mcDD$sp$class.class
scala/Function2$mcFJI$sp$class.class
scala/PartialFunction$OrElse.class
scala/Product19$class.class
scala/Enumeration$Value.class
```

As an added bonus, you can then probably support directory based classpath entries without writing additional code:

``` scala
scala> val dir = AbstractFile.getDirectory(new java.io.File("/code/scala/build/quick/classes/library"))
dir: scala.reflect.io.AbstractFile = PlainFile(PlainFile(), PlainFile(), PlainFile(PlainFile(PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile()), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(PlainFile(), PlainFile()), PlainFile(), PlainFile()), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), PlainFile(), Plain...


scala> flatten(dir).map(_.path).take(15).mkString("\n")
res6: String =
/code/scala/build/quick/classes/library/library.properties
/code/scala/build/quick/classes/library/rootdoc.txt
/code/scala/build/quick/classes/library/scala/annotation/Annotation.class
/code/scala/build/quick/classes/library/scala/annotation/bridge.class
/code/scala/build/quick/classes/library/scala/annotation/ClassfileAnnotation.class
/code/scala/build/quick/classes/library/scala/annotation/compileTimeOnly.class
/code/scala/build/quick/classes/library/scala/annotation/elidable$.class
/code/scala/build/quick/classes/library/scala/annotation/elidable.class
/code/scala/build/quick/classes/library/scala/annotation/implicitNotFound.class
/code/scala/build/quick/classes/library/scala/annotation/meta/beanGetter.class
/code/scala/build/quick/classes/library/scala/annotation/meta...
```

som-snytt commented on src/repl/scala/tools/nsc/interpreter/IMain.scala:
```diff
@@ -82,6 +86,9 @@ class IMain(@BeanProperty val factory: ScriptEngineFactory, initialSettings: Set
   private val _compiler: ReplGlobal                 = newCompiler(settings, reporter)   // our private compiler
 
+  private trait ExposeAddUrl extends URLClassLoader { def addNewUrl(url: URL) = this.addURL(url) }
+  private var _runtimeClassLoader: URLClassLoader with ExposeAddUrl = null              // wrapper exposing addURL
```
`ScalaClassLoader.URLClassLoader` also widens `addURL`.

heathermiller replied:
Good catch, fixed, thanks!

retronym commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -842,6 +842,147 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
+
+  /** Is given package class a system package class that cannot be invalidated?
+   */
+  private def isSystemPackageClass(pkg: Symbol) =
```
``` scala
private def isSystemPackageClass(pkg: Symbol) = pkg == RootClass || (pkg.hasTransOwner(ScalaPackageClass) && !pkg.hasTransOwner(ScalaToolsPackageClass)
```

Noting that:

```
scala> ScalaPackageClass.hasTransOwner(ScalaPackageClass)
res4: Boolean = true
```

heathermiller replied:
Done, thanks!

retronym commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -842,6 +842,147 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
+    pkg == RootClass ||
+    pkg == definitions.ScalaPackageClass || {
+      val pkgname = pkg.fullName
+      (pkgname startsWith "scala.") && !(pkgname startsWith "scala.tools")
```
Why the exception for `scala.tools`?

heathermiller replied:
This is actually logic left over from @gkossakowski's original `invalidateClasspathEntries` code. I’m not certain why `scala.tools is excepted`. I can remove this – but thought I was abiding by some inexplanable wisdom haha.

retronym commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -842,6 +842,147 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
+            if (elems.size == 1) elems.head
+            else new MergedClassPath(elems, classPath.context)
+          val oldEntries = mkClassPath(subst.keys)
+          val newEntries = mkClassPath(subst.values)
```
Is it a problem if that `oldEntries` and `newEntries` might have a in a non-deterministic order (given that `subst` is not a linked map)?

heathermiller replied:
I’m not sure, actually. To be safe, I just changed `subst` to be a `TreeMap` instead and defined an implicit ordering. So, this should force things to be in a deterministic order now.

heathermiller committed:
SI-6502 Addressing review comments

heathermiller commented:
RE: @retronym:

> I'd like to see this built on top of our existing abstractions for JAR file handling (`AbstractFile`).

Good idea. Done

Though it’ll still take a bit more work to be able to additionally deal with directories. This switch isn’t enough – I’m still having a ton of trouble reading in classes from dirs (for some reason, actual source is expected.)

heathermiller committed:
SI-6502 Addresses @som-snytt's feedback

heathermiller commented:
All feedback addressed, I believe.

som-snytt commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -612,13 +613,82 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+          exists = true
+        } catch {
+          case _: ClassNotFoundException => /* do nothing */
+        }
```
`intp.classLoader tryToLoadClass class.getName foreach (_ => exists = true)` is what I meant to suggest, or wait:

`if ((intp.classLoader tryToLoadClass class.getName).isDefined) exists = true`.

Did you know scalac came up with the idea for Seinfeld on tv?  It inferred a `Show[Nothing]`.

heathermiller replied:
That may be shorter, but is it clearer?

som-snytt replied:
I dunno.  I hate for convenience API to go unused.  Also, maybe from Effective Java, and from paulp dogma, two funcs foo and notfoo are preferred to f(foo = true), or respectively, Foo and NotFoo with f implementations.  But to clarify, recently I said somewhere that using guava's `ImmutableMap.of()` is not clearer than `Collections.emptyMap()`, just because everyone knows the core API.

In general, though, an API that reduces braces is a win.  Here, I have to wonder if catching CNFE suffices.  The ScalaCL also catches SecurityEx.  Should it?  I hope so, but in any case I don't want to think about that right now.

Off the record, I also said to myself, I thought I just looked at loading a class from a byte array, do we really need to introduce a class for that?  But I didn't have a 30-second response.  That's why it's nice to promote existing enhancements like ScalaClassLoader.  But I don't always prefer DRY over quick'n'dirty re-implementations that avoid extraneous dependencies.

Sorry for the tldr, it's a long Hulu commercial during once upon a time.

heathermiller replied:
Ge-done-t. :)

heathermiller committed:
SI-6502 Addresses comments by @som-snytt

retronym commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -612,13 +613,76 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+        val clazz = cloader.classOf(arr)
+        if ((intp.classLoader tryToLoadClass clazz.getName).isDefined) exists = true
+      }
+    }
```
This section can still do with some polish.

``` scala
def classNameOf(classFile: AbstractFile): String = cloader.classOf(classFile.toByteArray).getName
def alreadyDefined(clsName: String) = intp.classLoader.tryToLoadClass(clsName).isDefined
val exists = entries.filter(_.hasExtension("class")).map(classNameOf).exists(alreadyDefined)
```

You can then delete `readFully`.

retronym commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -612,13 +613,76 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+    } else if (exists) {
+      echo("The path '" + f + "' cannot be loaded, because existing classpath entries conflict.")
+    } else echo("The path '" + f + "' doesn't seem to exist.")
+  }
```
This conditional seems awkwardly arranged.

How about:

``` scala
if (!f.exists) echo(s"The path '$f' doesn't seem to exist.")
else if (exists) echo(s"The path '$f' cannot be loaded, because existing classpath entries conflict.") // TODO tell me which one
else ...
```

retronym commented:
How has this been tested? Are we able to add automated tests?

heathermiller commented:
Manually, where I take a bunch of sometimes overlapping jars and load them in different ways and try to use them. If you can suggest a way for me to automatically test arbitrary jar loading (I add random jars to the test suite...? Not sure that makes sense) I'd be happy to add a test case.

retronym commented:
Sometime manual testing is okay, but you should include a transcript of the test process in the pull request.

We do have a mechanism for using JARs in tests:

```
% find test  | grep desired
test/benchmarks/lib/jsr166_and_extra.jar.desired.sha1
test/files/codelib/code.jar.desired.sha1
test/files/lib/annotations.jar.desired.sha1
test/files/lib/enums.jar.desired.sha1
test/files/lib/genericNest.jar.desired.sha1
test/files/lib/jsoup-1.3.1.jar.desired.sha1
test/files/lib/macro210.jar.desired.sha1
test/files/lib/methvsfield.jar.desired.sha1
test/files/lib/nest.jar.desired.sha1
test/files/speclib/instrumented.jar.desired.sha1
```

One needs authorization to upload the JARs into the repository (in Github we just have the signature, and the ANT build downloads them). I haven't personally done this before and don't have the credentials.

I know that @xeno-by has done this, though. For example, in b10f45a78217b002f8ac6e2051ff932a1ac2e029, he depends on JAR in a new test case.

An alternative, which may be better in this case, is to use a `run` test case to programmatically generate the JAR files. Take a look at test/files/run/t6440b.scala for an example of compiling a few batches of code. You could use `s"-d ${testOutput.toFile.getPath}/jar1.jar"` as the options to a global instance to emit some code to a JAR.

Once the JARs were prepared, you could programatically drive the interpreter.

I know this represents another half a day of hacking about, but this feature seems subtle enough to warrant that.

xeno-by commented:
Can tell you guys how to do this stuff. We could do this via email.

mpociecha commented:
@heathermiller Do you mean certain concrete jars? Or maybe is only this overlapping important? Theoretically you could generate some jars (see https://github.com/mpociecha/scala/blob/flat-classpath/test/files/run/various-flat-classpath-types.scala).

heathermiller committed:
Adds tests that programatically generate jars

heathermiller committed:
Addresses review comments

heathermiller commented:
Hey guys, thanks for all the pointers. I think all comments are now addressed (clean ups as suggested by @retronym applied, and 4 tests that programmatically generate jars are now included)

retronym commented:
LGTM!

heathermiller commented:
woot! :) thanks for the time you spent on this, Jason!

retronym commented:
We'll hold off on merging a few days until we officially release 2.11.4 and open this branch for 2.11.5 development.

heathermiller commented:
Ah, I realized I haven't squashed this yet. Do your scripts do the squash? (Asking because that's the databricks workflow – all PRs get squashed into one commit and that commit is built/tested against)

If not, no worries, I can squash when I'm not on Bart :)

retronym commented:
We don't auto squash. You can leave as many commits in the history as makes sense to tell the story of the change. I don't mind the extra commits with refinements from review comments in this case.

lrytz commented:
Ping @heathermiller – the failure in last commit looks legit (https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/7398/console)

```
quick.comp:
    [mkdir] Created dir: /localhome/jenkins/c/workspace/pr-scala-test/scala/build/quick/classes/compiler
[quick.compiler] Compiling 301 files to /localhome/jenkins/c/workspace/pr-scala-test/scala/build/quick/classes/compiler
[quick.compiler] /localhome/jenkins/c/workspace/pr-scala-test/scala/src/compiler/scala/tools/nsc/Global.scala:878: error: not found: type MergedClassPath
[quick.compiler]       case cp: MergedClassPath[_] =>
[quick.compiler]                ^
```

heathermiller commented:
Well, I invested a fair amount of time trying, and I can't reproduce the failure. MergedClassPath is imported: https://github.com/heathermiller/scala/blob/repl-cp-fix2/src/compiler/scala/tools/nsc/Global.scala#L17

It actually looks like Jenkins is building the wrong thing. Because the error reported by jenkins says that on line 878 "case cp: MergedClassPath[_] =>" which is _not_ the case. This case is actually on 879.

Also, the third to last commit was built successfully, containing the same Global.scala.

lrytz commented:
You're unlucky, it seems to be a merge conflict that's not caught by git. I can reproduce it locally:

```
git co ae70f020cdd22d155882191ca60d40542615b606 // part of current 2.11.x
git merge 3513dd53648b75c32877efa4d1e924d63830e03e // your commit, no merge failure
ant -Dlocker.skip=1
[...]
quick.comp:
[quick.compiler] Compiling 25 files to /Users/luc/scala/scala/build/quick/classes/compiler
[quick.compiler] /Users/luc/scala/scala/src/compiler/scala/tools/nsc/Global.scala:878: error: not found: type MergedClassPath
[quick.compiler]       case cp: MergedClassPath[_] =>
[quick.compiler]                ^
```

Try merging / rebasing on current 2.11.x.

heathermiller commented:
Hopefully now ok :)

heathermiller committed:
SI-6502 Moving methods concerned with the state of Global from IMain to Global

heathermiller commented:
Hey, @retronym and @lrytz, I just added a commit that shuffles things around a bit.

I think it would be better to move `extendCompilerClassPath` and `mergeUrlsIntoClassPath` into `Global`, since they only depend on, and update, the state of `global`. They are completely independent of `IMain`. This would also allow Spark's REPL to make use of this fix without being forced to put classes into the `scala.tools.nsc` package, or to use reflection, for instance, to access `platform.currentClassPath`, or, worse, to do both, and duplicate a lot of code from `IMain` (that’s what’s currently required).

This seems sensible/innocuous to me – could you let me know if you see any issues with this?

retronym commented:
That sounds okay to me. But perhaps we could review and discuss that as followup pull request?

heathermiller commented:
I can break it out if you'd like – though it complicates this companion/concurrent PR to Spark that depends on this PR being merged :/

retronym commented:
No problems, feel free to add it as an extra commit to this PR.

Looking more closely at `mergeUrlsIntoClassPath`, I think there are a few improvements.
- It only uses `platform.classpath`, how about it a member on `ClassPath` to make it usable without a `JavaPlatform`?
- The if/else/else to create the right sort of `AbstractFile` looks like it should be factored out to a new factory method on `AbstractFile`. `AbstractFile.getURL` is currently unused within the compiler codebase, and might be a good candidate to generalize/document to meet the needs.

heathermiller committed:
SI-6502 Refactorings suggested by review
- Moves mergeUrlsIntoClassPath from Global into ClassPath
- Revises and documents AbstractFile.getURL

heathermiller commented:
All good ideas! I just addressed your feedback and added a new commit with the suggested refactorings. Let me know if anything looks amiss!

Thanks :)

gkossakowski commented:
Hey guys! I'm a bit late to the party (my Mac was down for almost 2 weeks) but I'll try to give my feedback in next 2-3 days.

retronym commented:
LGTM

A note for next time: It would be easier to review if that were two commits: one that refactored the method in-situ, and a second that moved it. That way we can see the diff more clearly.

retronym commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 62669006)
:cat: Roger! Rebuilding pr-scala for 81925718, a84abd08, f65c4309, 04ee526c, d045ddeb, be3eb58d, 9e56c7a6, 24a2ef97. :rotating_light:

heathermiller commented:
True, good point. Duly noted :)

heathermiller commented:
Thanks again for the detailed review!

heathermiller commented:
Ping on this – just wondering when this will be merged? I ask because this PR is blocking Spark from moving to 2.11 (as things stood when I was in CA 1.5 weeks ago).

heathermiller commented:
The sooner we get a SNAPSHOT or something, the better – gives us more time to have the new REPL support in the Spark 1.2 branch and more time to have this tested before 1.2 is released.

retronym commented:
@gkossakowski Do you still plan to review this? If you don't have time in the next day or two, I propose we merge this as is.

gkossakowski commented:
Unfortunately, I won't have time for detailed review. The infra work and dealing with backlog is consuming all of my time. I trust review powers of @retronym, though. :)

There's only one thing that concerns me: we are adding a bunch of logic back to `Global`. I know originally the classpath invalidation logic was in Global but that was a mistake. I think this logic should live in a separate class with properly spelled dependencies. I'd propose moving this logic into a class that is referenced from Global through composition. If Spark expects certain methods to live directly in Global, let's just add deprecated forwarders. This would be part of the quest for reducing Global's surface area.

The refactoring work can be done in separate PR (there's no need to block this one any further!) but I'd to see it happen. @heathermiller, WDYT?

gkossakowski commented:
See also #4060 that is directly related to this work. It would benefit from clearly delineating the boundary between classpath invalidation and the rest of the compiler.

heathermiller commented:
Maybe later, sure, but this PR has been blocking me for over a month now. Would be nice to merge it...

heathermiller commented:
Thanks very much!! :)
